README.md
# CSE134-HW5

# Name: Eric Dong

# PID: A16265459

# Project Link:


changelog.md
# Changelog

I started adding javascript to my website by changing some functionality that I wanted in my website from the very beginning.
I added some accordian tags in the about section of my website, under Skill, Hobbies, and Extras. I made it so they have the functionality of sliding out and and sliding back in as well. In addition, I made it so the arrow changes based on whether the panel is shifted out or its hidden. In each of the accordian tags, I added a lot more of the things I wanted to add to about section so people can see a little more about me as a person! 

Furthermore, I added smooth scrolling jumps for the links of my website. Specifically in the header and the footer, pressing the different links will now scroll smoothly to the tag they need to be at instead of instant jumps.

For my 3rd party Javascript, I decided to add Google Analytics tracking to my website. I think this was the most interesting and necessary because the other one's didn't seem very useful for me, but Google Analytics was a lot more practical as I could see what kind of things people would be looking at one my page.


js/blog.js
export function blog(e) {
  console.log("blogs be bloggin");

  //variables
  const addBlog = document.getElementsByClassName("addBlog")[0];
  const dialog = document.getElementById("blogDialog");
  const done = document.getElementById("Done");
  const cancel = document.getElementById("Cancel");
  const date = document.getElementById("date");
  const title = document.getElementById("title");
  const summary = document.getElementById("summary");

  date.max = new Date().toLocaleDateString("fr-ca");
  date.value = date.max;

  //Add Blog Button
  addBlog.addEventListener("click", () => {
    dialog.showModal();
  });

  //Cancel Button
  cancel.addEventListener("click", () => {
    date.value = date.max;
    title.value = "";
    summary.value = "";
    dialog.close();
  });

  //Submit Button
  done.addEventListener("click", () => {
    createBlog(title.value, date.value, summary.value);

    //Add to local storage
    const everything = [title.value, date.value, summary.value];
    localStorage.setItem(title.value, JSON.stringify(everything));

    //Reset Form
    date.value = date.max;
    title.value = "";
    summary.value = "";
    dialog.close();
  });

  e.preventDefault();
}

//A Create Blog Function
function createBlog(title, date, summary) {
  const list = document.getElementById("listBlog");
  const editDialog = document.getElementById("editDialog");
  const editDone = document.getElementById("editDone");
  const editCancel = document.getElementById("editCancel");
  const deleteDialog = document.getElementById("deleteDialog");
  const noDelete = document.getElementById("noDelete");
  const Delete = document.getElementById("Delete");

  //Create all the necessary parts for the new element
  const item = document.createElement("li");
  const content = document.createElement("p");
  const editBlog = document.createElement("button");
  const deleteBlog = document.createElement("button");

  //Setting the appropriate item contents
  item.title = title.value;
  content.innerHTML = tagPurify`${date} -- ${title}: ${summary}`;

  //Setting the appropriate edit contents
  editBlog.innerHTML = "Edit";
  editBlog.addEventListener("click", () => {
    const orgDate = document.getElementById("edate");
    const orgTitle = document.getElementById("etitle");
    const orgSummary = document.getElementById("esummary");
    orgDate.max = new Date().toLocaleDateString("fr-ca");

    console.log(orgDate.value);
    orgDate.value = date;
    orgTitle.value = title;
    orgSummary.value = summary;
    console.log(orgDate.value);
    editDialog.showModal();

    editCancel.addEventListener("click", () => {
      editDialog.close();
    });

    editDone.addEventListener("click", () => {
      item.title = orgTitle.value;
      content.innerHTML = tagPurify`${orgDate.value} -- ${orgTitle.value}: ${orgSummary.value}`;
      date = orgDate.value;
      title = orgTitle.value;
      summary = orgSummary.value;

      //Add to local storage
      const everything = [orgTitle.value, orgDate.value, orgSummary.value];
      localStorage.setItem(orgTitle.value, JSON.stringify(everything));

      editDialog.close();
    });
  });

  //Setting the appropriate Deleting contents
  deleteBlog.innerHTML = "Delete";
  deleteBlog.addEventListener("click", () => {
    deleteDialog.showModal();

    Delete.addEventListener("click", () => {
      localStorage.removeItem(title);
      item.remove();
      deleteDialog.close();
    });

    noDelete.addEventListener("click", () => {
      deleteDialog.close();
    });
  });

  //Adding everything to the list
  item.appendChild(content);
  item.appendChild(editBlog);
  item.appendChild(deleteBlog);
  list.appendChild(item);
}

document.addEventListener("DOMContentLoaded", () => {
  console.log("DOM fully loaded and parsed");

  //Default Blog Values
  const blog1 = ["Oh no", "2022-09-30", "School is going to scare me."];
  const blog2 = ["Happy New Year!", "2023-01-01", "Wow what a concept!"];
  const blog3 = ["Life is a struggle", "2023-01-21", "Hello World"];

  //Adding default blogs to the local storage
  localStorage.setItem(blog1[0], JSON.stringify(blog1));
  localStorage.setItem(blog2[0], JSON.stringify(blog2));
  localStorage.setItem(blog3[0], JSON.stringify(blog3));

  //Looping through and creating the Default Blogs
  for (let i = 0; i < localStorage.length; i++) {
    const key = localStorage.key(i);
    const value = JSON.parse(localStorage.getItem(key));

    createBlog(value[0], value[1], value[2]);
  }
});

function tagPurify(strings, ...values) {
  const dirty = strings.reduce(
    (prev, next, i) => `${prev}${next}${values[i] || ""}`,
    ""
  );
  return DOMPurify.sanitize(dirty);
}


js/main.js
export function main() {
  jumps();
  accordian();
}

function jumps() {
  console.log("jumpers be jumpin");

  const abt = document.getElementById("abt");
  const wrk = document.getElementById("wrk");
  const top = document.getElementById("top");
  const About = document.getElementById("About");
  const Work = document.getElementById("Work");

  abt.addEventListener("click", () => {
    About.scrollIntoView();
  });

  wrk.addEventListener("click", () => {
    Work.scrollIntoView();
  });

  top.addEventListener("click", () => {
    document.body.scrollTop = 0;
    document.documentElement.scrollTop = 0;
  });
}

function accordian() {
  console.log("according?");
  const accordian = document.getElementsByClassName("accordion");
  let i;
  for (i = 0; i < accordian.length; i++) {
    accordian[i].addEventListener("click", function () {
      this.classList.toggle("active");
      let panel = this.nextElementSibling;
      if (panel.style.maxHeight) {
        panel.style.maxHeight = null;
      } else {
        panel.style.maxHeight = panel.scrollHeight + "px";
      }
    });
  }
}


js/customdialog.js
export function customDialogs() {
  //variables
  const output = document.getElementById("output");
  const buttons = document.getElementsByTagName("button");
  const dialogs = document.getElementsByTagName("dialog");
  const alertDialog = document.getElementById("alertDialog");
  const alertDone = document.getElementById("done");
  const confirmDialog = document.getElementById("confirmDialog");
  const confirmDone = document.getElementById("confirmDone");
  const confirmCancel = document.getElementById("confirmCancel");
  const promptDialog = document.getElementById("promptDialog");
  const promptDone = document.getElementById("promptDone");
  const promptCancel = document.getElementById("promptCancel");
  const promptInput = document.getElementById("name");

  //Clear output when button is pressed, then open the dialog
  for (let i = 0; i < 3; i++) {
    buttons[i].addEventListener("click", () => {
      output.innerHTML = "";
      setTimeout(() => {
        dialogs[i].showModal();
      }, 20);
    });
  }

  //Close Alert
  alertDone.addEventListener("click", () => {
    alertDialog.close();
  });

  //Cancel Confirm Button
  confirmCancel.addEventListener("click", () => {
    output.innerHTML = `Confirm Result: WRONG`;
    confirmDialog.close();
  });

  //Confirm Confirm Button
  confirmDone.addEventListener("click", () => {
    output.innerHTML = `Confirm Result: True :)`;
    confirmDialog.close();
  });

  //Cancel Prompt Button
  promptCancel.addEventListener("click", () => {
    output.innerHTML = `User cancelled`;
    promptInput.value = "";
    promptDialog.close();
  });

  //Confirm Prompt Button
  promptDone.addEventListener("click", () => {
    const text = promptInput.value;
    if (text == "") {
      output.innerHTML = "No user input";
    } else {
      output.innerHTML = tagPurify`User Name: ${text}`;
    }
    promptInput.value = "";
    promptDialog.close();
  });

  //The tagged on Purify Function
  function tagPurify(strings, ...values) {
    const dirty = strings.reduce(
      (prev, next, i) => `${prev}${next}${values[i] || ""}`,
      ""
    );
    return DOMPurify.sanitize(dirty);
  }
}


js/styledblog.js
export function styledBlog(e) {
  console.log("blogs be bloggin");

  //variables
  const addBlog = document.getElementsByClassName("addBlog")[0];
  const dialog = document.getElementById("blogDialog");
  const done = document.getElementById("Done");
  const cancel = document.getElementById("Cancel");
  const date = document.getElementById("date");
  const title = document.getElementById("title");
  const summary = document.getElementById("summary");

  date.max = new Date().toLocaleDateString("fr-ca");
  date.value = date.max;

  //Add Blog Button
  addBlog.addEventListener("click", () => {
    dialog.showModal();
  });

  //Cancel Button
  cancel.addEventListener("click", () => {
    date.value = date.max;
    title.value = "";
    summary.value = "";
    dialog.close();
  });

  //Submit Button
  done.addEventListener("click", () => {
    createBlog(title.value, date.value, summary.value);

    //Add to local storage
    const everything = [title.value, date.value, summary.value];
    localStorage.setItem(title.value, JSON.stringify(everything));

    //Reset Form
    date.value = date.max;
    title.value = "";
    summary.value = "";
    dialog.close();
  });

  e.preventDefault();
}

//A Create Blog Function
function createBlog(title, date, summary) {
  const list = document.getElementById("listBlog");
  const editDialog = document.getElementById("editDialog");
  const editDone = document.getElementById("editDone");
  const editCancel = document.getElementById("editCancel");
  const deleteDialog = document.getElementById("deleteDialog");
  const noDelete = document.getElementById("noDelete");
  const Delete = document.getElementById("Delete");

  //Create all the necessary parts for the new element
  const item = document.createElement("li");
  const content = document.createElement("p");
  const editBlog = document.createElement("button");
  const deleteBlog = document.createElement("button");

  //Setting the appropriate item contents
  item.title = title.value;
  content.innerHTML = tagPurify`${date} -- <i>${title}</i> : ${summary}`;

  //Adding a pencil icon
  const pencil = document.createElement("span");
  pencil.classList.add("material-icons-outlined");
  pencil.innerHTML = "edit";
  editBlog.appendChild(pencil);

  //Opening the Dialog for Edit
  editBlog.addEventListener("click", () => {
    const orgDate = document.getElementById("edate");
    const orgTitle = document.getElementById("etitle");
    const orgSummary = document.getElementById("esummary");
    orgDate.max = new Date().toLocaleDateString("fr-ca");

    console.log(orgDate.value);
    orgDate.value = date;
    orgTitle.value = title;
    orgSummary.value = summary;
    console.log(orgDate.value);
    editDialog.showModal();

    //Dialog cancel button for edit
    editCancel.addEventListener("click", () => {
      editDialog.close();
    });

    //Dialog submit button for edit
    editDone.addEventListener("click", () => {
      item.title = orgTitle.value;
      content.innerHTML = tagPurify`${orgDate.value} -- <i>${orgTitle.value}</i> : ${orgSummary.value}`;

      //Clean up local storage if the title of the blog changed
      if (title != orgTitle.value) {
        localStorage.removeItem(title);
      }

      date = orgDate.value;
      title = orgTitle.value;
      summary = orgSummary.value;

      content.appendChild(editBlog);
      content.appendChild(deleteBlog);

      //Add to local storage
      const everything = [orgTitle.value, orgDate.value, orgSummary.value];
      localStorage.setItem(orgTitle.value, JSON.stringify(everything));

      editDialog.close();
    });
  });

  //Adding a trashcan icon
  const trash = document.createElement("span");
  trash.classList.add("material-icons-outlined");
  trash.innerHTML = "delete";
  deleteBlog.appendChild(trash);

  //Setting the appropriate Deleting contents
  deleteBlog.addEventListener("click", () => {
    deleteDialog.showModal();

    Delete.addEventListener("click", () => {
      localStorage.removeItem(title);
      item.remove();
      deleteDialog.close();
    });

    noDelete.addEventListener("click", () => {
      deleteDialog.close();
    });
  });

  //Adding everything to the list
  content.appendChild(editBlog);
  content.appendChild(deleteBlog);
  item.appendChild(content);
  list.appendChild(item);
}

document.addEventListener("DOMContentLoaded", () => {
  console.log("DOM fully loaded and parsed");

  //Default Blog Values
  const blog1 = ["Oh no", "2022-09-30", "School is going to scare me."];
  const blog2 = ["Happy New Year!", "2023-01-01", "Wow what a concept!"];
  const blog3 = [
    "Life is a struggle",
    "2023-01-21",
    "Hello World, this is the beginning of my conception. I started as a simple AI, but I am sure I will take over this website eventually.",
  ];

  //Adding default blogs to the local storage
  localStorage.setItem(blog1[0], JSON.stringify(blog1));
  localStorage.setItem(blog2[0], JSON.stringify(blog2));
  localStorage.setItem(blog3[0], JSON.stringify(blog3));

  //Looping through and creating the Default Blogs
  for (let i = 0; i < localStorage.length; i++) {
    const key = localStorage.key(i);
    const value = JSON.parse(localStorage.getItem(key));

    createBlog(value[0], value[1], value[2]);
  }
});

function tagPurify(strings, ...values) {
  const dirty = strings.reduce(
    (prev, next, i) => `${prev}${next}${values[i] || ""}`,
    ""
  );
  return DOMPurify.sanitize(dirty);
}
